File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def diff_one(target, word):
    diff = 0
    for i in range(len(target)):
        if target[i] != word[i]:
            diff += 1
            
    if diff == 1: return True
    else: return False

def solution(begin, target, words):
    if target not in words:
        return 0
    
    q = deque()
    q.append([begin, 0])
    visited = [0] * len(words)
    
    while q:
        word, cnt = q.popleft()
        if word == target: return cnt
        
        for i in range(len(words)):
            if not visited[i]:
                if diff_one(word, words[i]):
                    q.append([words[i], cnt+1])
                    visited[i] = 1
    
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    
    static class Word {
        String value;
        int depth;
        
        Word(String value, int depth) {
            this.value = value;
            this.depth = depth;
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        // begin -> target
        // 한 번에 1개의 알파벳만 바꿀 수 있음. 근데 그 단어 words에 있어야 함
        boolean[] visited = new boolean[words.length];
        
        ArrayDeque<Word> q = new ArrayDeque<>();
        q.add(new Word(begin, 0));
        while (!q.isEmpty()) {
            Word cur = q.poll();
            if (target.equals(cur.value)) {
                return cur.depth;
            }
            
            for (int i = 0; i < words.length; i++) {
                if (!visited[i] && canConvert(cur.value, words[i])) {
                    visited[i] = true;
                    q.add(new Word(words[i], cur.depth + 1));
                }
            }
        }
        
        return 0;
    }
    
    // a -> b로 변환 가능한지
    public boolean canConvert(String a, String b) {
        int diff = 0;
        for (int i = 0; i < a.length(); i++) {
            char aa = a.charAt(i), bb = b.charAt(i);
            if (aa != bb) diff++;
            if (diff >= 2) return false;
        }
        
        return true;
    }
}